lib/client.js

```diff
@@ -50,28 +49,98 @@ class CloudantClient {
     // build plugin array
     var plugins = [];
     if (typeof this._cfg.plugin === 'undefined' && typeof this._cfg.plugins === 'undefined') {
-      plugins = ['cookieauth']; // default
+      plugins = [ 'cookieauth' ]; // default
     } else {
       [].concat(self._cfg.plugin).concat(self._cfg.plugins).forEach(function(plugin) {
-        if (typeof plugin === 'function' && !(plugin._pluginVersion >= 2)) {
-          if (self.useLegacyPlugin) {
-            throw new Error('Using multiple legacy plugins in not supported');
-          } else {
-            debug('Using legacy plugin.');
-            self.useLegacyPlugin = true;
-            client = plugin; // use legacy plugin as client
-            return;
-          }
+        if (typeof plugin !== 'function' || plugin.pluginVersion >= 2) {
+          plugins.push(plugin);
+        } else if (self.useLegacyPlugin) {
+          throw new Error('Using multiple legacy plugins in not supported');
+        } else {
+          self.useLegacyPlugin = true;
+          client = plugin; // use legacy plugin as client
         }
-        plugins.push(plugin);
       });
     }
 
     // initialize the internal client
     self._initClient(client);
 
     // add plugins
-    self.addPlugins(plugins);
+    self._addPlugins(plugins);
+  }
+
+  _addPlugins(plugins) {
+    var self = this;
+
+    if (!Array.isArray(plugins)) {
+      plugins = [ plugins ];
+    }
+
+    var defaultCfg = {
+      retryDelayMultiplier: DEFAULTS.retryDelayMultiplier,
+      retryInitialDelayMsecs: DEFAULTS.retryInitialDelayMsecs
+    };
+
+    plugins.forEach(function(p) {
+      if (typeof p === 'undefined') {
+        return;
+      }
+
+      var cfg, Plugin;
+
+      if (typeof p === 'function') {
+        debug(`Found custom plugin: '${p.id}'`);
+        Plugin = p;
+        cfg = {};
+      } else {
+        var plugin = {};
+
+        // expected: { 'pluginName': { 'configKey1': 'configValue1', ... } }
+        if (typeof p !== 'object') {
```

Comment:
At the moment this assumes a string for non-object, but what if someone accidentally specifies `1` or `true` as a plugin?

Reply:
It will throw `Failed to load plugin`.

Added some additional tests in 4c2b4503de1728ff5b5cda03e97eca9bbb4931ab.